util/topsql/stmtstats/stmtstats.go

```diff
@@ -16,30 +16,75 @@ package stmtstats
 
 import (
 	"sync"
+	"time"
 
 	"go.uber.org/atomic"
 )
 
 var _ StatementObserver = &StatementStats{}
 
+// nowFunc returns the current time; it's overridden in tests.
+var nowFunc = time.Now
+
 // StatementObserver is an abstract interface as a callback to the corresponding
 // position of TiDB's SQL statement execution process. StatementStats implements
 // StatementObserver and performs counting such as SQLExecCount/SQLDuration internally.
 // The caller only needs to be responsible for calling different methods at the
 // corresponding locations, without paying attention to implementation details.
 type StatementObserver interface {
-	// OnExecutionBegin should be called before statement execution.
-	OnExecutionBegin(sqlDigest, planDigest []byte)
+	// OnDispatchBegin should be called on the begin of TiDB dispatch command.
+	OnDispatchBegin()
+
+	// OnDispatchFinish should be called on TiDB finished dispatch command.
+	OnDispatchFinish()
+
+	// OnHandleQueryBegin should be called on the begin of TiDB handle query.
+	OnHandleQueryBegin()
+
+	// OnHandleQueryFinish should be called on TiDB finished handle query.
+	OnHandleQueryFinish()
+
+	// OnHandleStmtBegin should be called on the begin of TiDB handle stmt.
+	OnHandleStmtBegin()
+
+	// OnHandleStmtFinish should be called on TiDB finished handle stmt.
+	OnHandleStmtFinish()
+
+	// OnStmtReadyToExecute should be called on statement ready to execute with physical plan.
+	// It's ok to call this function multiple time, but only the first time take effect. This is design for execution retry.
+	OnStmtReadyToExecute(sqlDigest, planDigest []byte)
+
+	// OnHandleStmtBegin should be called on the begin of TiDB handle stmt.
+	OnHandleStmtExecuteBegin()
+
+	// OnHandleStmtFinish should be called on TiDB finished stmt execute.
+	OnHandleStmtExecuteFinish()
+
+	// OnHandleStmtBegin should be called on the begin of TiDB handle stmt fetch.
+	OnHandleStmtFetchBegin()
 
-	// OnExecutionFinished should be called after the statement is executed.
-	OnExecutionFinished(sqlDigest, planDigest []byte)
+	// OnHandleStmtFinish should be called on TiDB finished stmt fetch.
+	OnHandleStmtFetchFinish(sqlDigest, planDigest []byte)
+
+	// OnHandleStmtBegin should be called on the begin of TiDB use db.
+	OnUseDBBegin()
+
+	// OnHandleStmtFinish should be called on TiDB finished use db.
+	OnUseDBFinish()
+
+	// OnHandleInternalStmtBegin should be called on the begin of TiDB handle internal stmt.
+	OnHandleInternalStmtBegin()
+
+	// OnHandleInternalStmtFinish should be called on TiDB finished handle internal stmt.
+	OnHandleInternalStmtFinish()
```

Comment:
Why do we need an internal statement? What will happen if it is not processed? Will we miss background SQL calls?